File: 1038-binary-search-tree-to-greater-sum-tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        
        self.sum = 0
        
        def reverse_inorder_traversal(node):
            if node:
                reverse_inorder_traversal(node.right)
                self.sum += node.val
                node.val = self.sum
                reverse_inorder_traversal(node.left)
        
        reverse_inorder_traversal(root)
        return root

File: 1038-binary-search-tree-to-greater-sum-tree/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func bstToGst(root *TreeNode) *TreeNode {
    
    reverseInorderTraversal(root, 0)
    return root
}


func reverseInorderTraversal(node *TreeNode, sum int) int {
    if node != nil {
        sum = reverseInorderTraversal(node.Right, sum)
        sum += node.Val
        node.Val = sum
        sum = reverseInorderTraversal(node.Left, sum)
    }
    return sum
}
